File: 421.maximum-xor-of-two-numbers-in-an-array.py
#
# @lc app=leetcode id=421 lang=python3
#
# [421] Maximum XOR of Two Numbers in an Array
#

# @lc code=start
from typing import List


class Solution:
    '''
    As LC made testcase to TLE all the python Trie solutions, we resort to something else. A simpler way to do it is to consider the different bits. Going from large to small, we <<= 1 each level and ask if ans + 1 exist in the set of prefixes.
    '''
    def findMaximumXOR(self, nums: List[int]) -> int:
        result = 0
        numberOfLevels = max(nums).bit_length()
        for i in range(numberOfLevels - 1, -1, -1):
            prefixes = set([num >> i for num in nums])
            result <<= 1
            for prefix in prefixes:
                if (result + 1) ^ prefix in prefixes:
                    result += 1
                    break
        return result
# @lc code=end

File: 421.maximum-xor-of-two-numbers-in-an-array.cpp
/*
 * @lc app=leetcode id=421 lang=cpp
 *
 * [421] Maximum XOR of Two Numbers in an Array
 */

// @lc code=start
#include <vector>
#include <unordered_set>
#include <algorithm>
#include <cmath>

class Solution {
public:
    int findMaximumXOR(std::vector<int>& nums) 
    {
        int result = 0;
        int numberOfLevels = 31;

        for (int i = numberOfLevels - 1; i >= 0; i--) 
        {
            std::unordered_set<int> prefixes;

            for (int num : nums) 
            {
                prefixes.insert(num >> i);
            }

            result <<= 1;

            for (int prefix : prefixes) 
            {
                if (prefixes.count((result + 1) ^ prefix)) 
                {
                    result += 1;
                    break;
                }
            }
        }

        return result;
    }
};

// @lc code=end

